3595 3766 adds a base writer and an itk writer

wyli opened the pull request:
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

part of #3595

### Description
this PR adds a writer base class and an itkwriter subclass,
internally the affines are assumed in the `RAS` space to be aligned with nifti.

more consistency tests will be added when the nibabelwrite is in place.
the tests are ready in https://github.com/Project-MONAI/MONAI/pull/3443

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

wyli commented:
/build

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        """
+        # change data to "channel last" format
+        if channel_dim is not None:
+            # _chns = ensure_tuple_rep(channel_dim)
```
please remove the comment code.

Thanks.

wyli replied:
Yes, I forgot to reenable the multiple channel_dim here, will update in the latest

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        Args:
+            data: input data to be converted to "channel-last" format.
+            channel_dim: specifies the axis of the data array that is the channel dimension.
+                ``None`` indicates no channel dimension.
```
Could you please help describe the supported values? I remember it can be multiple channels dims now?
I checked below code, seems it only supports 1 channel dim: `while len(data.shape) < spatial_ndim + 1`.

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+            data = moveaxis(data, channel_dim, -1)
+        else:  # adds a channel dimension
+            data = data[..., None]
+        # To ensure at least three spatial dims
```
Seems the code can support not 3 dims?

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        return data_np, ensure_mat44(target_affine)
+
+    @classmethod
+    def convert_to_channel_last(
```
I feel it's better to change this logic to the transform `AsChannelLast`, then users can apply below 3 transforms in postprocessing:
```py
AffineResample()  # the resample logic in this PR
AsChannelLast()
SaveImage()
```
What do you think?
@ericspod @rijobro for more discussions.

Thanks.

wyli replied:
The current design is the other way around: `SaveImage` as a transform should work well with the channel-first format, so that we hide the backend implementation details. For example, we can `SaveImage` to 'dump' intermediate transform outcomes conveniently. `SaveImage` internally calls resample/channel_last steps based on the requirements of the writer backend, not all backend consume the same channel last format.

Nic-Ma replied:
Hi @wyli ,

Thanks for your explanation, I see your design point.
But I am not very sure which way is better to go, as one of our MONAI design principles is "easy to integrate", maybe we should keep every component simple and independent? Daguang also raised the idea to extract `resample` from `SaveImage` during Clara 4.0 development.
@ericspod @rijobro Could you guys please help share your opinions?

Thanks in advance.

wyli replied:
I'm trying to create a `SpatialResample` transform, and `SaveImage` will internally call this transform if the user runs `SaveImage(resample=True)`， the user would also have the `SpatialResample` transform. I don't see why `SaveImage` shouldn't integrate the resampling functionality.

Nic-Ma replied:
Hi @wyli ,

That makes sense to me, if also providing `SpatialResample` transform, it's OK to call it in `SaveImage`.
And so this `resample` logic would be in the `SpatialResample` transform instead of this `ImageWriter` module, right?

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        """
+        if not metadata:
+            default_dict = {"original_affine": None, "affine": None, "spatial_shape": None}
+            metadata = default_dict
```
Can be simplified: `metadata = {"original_affine": None, "affine": None, "spatial_shape": None}`
And seems no need to set default with `None`, just `metadata = {}` is same?

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        """the constructor supports adding new instance members."""
+        self.data_obj = None
+        for k, v in kwargs.items():
+            setattr(self, k, v)
```
I didn't find the usage of these attributes.
And could you please help add doc-string to describe what's the expected values of this `kwargs` arg?

Thanks.

wyli replied:
After your initial comment, I've updated the docstring:
```
The constructor supports adding new instance members.
The current member in the base class is ``self.data_obj``, the subclasses can add more members,
so that necessary meta information can be stored in the object and shared among the class methods.
```

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+
+
+@require_pkg(pkg_name="itk")
+class ITKWriter(ImageWriter):
```
Could you please help add doc-strings of **args** for all the functions in this class?

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+            affine=affine,
+            target_affine=original_affine if resample else None,
+            output_spatial_shape=spatial_shape,
+            mode=options.pop("mode", GridSampleMode.BILINEAR),
```
I feel it's necessary to describe all the supported args in `options` in the doc-string.

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+            data=data_array,
+            channel_dim=channel_dim,
+            squeeze_end_dims=squeeze_end_dims,
+            spatial_ndim=kwargs.pop("spatial_ndim", 3),
```
I feel it's necessary to describe all the supported args in `kwargs` in the doc-string.

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        data_array = super().create_backend_obj(data_array)
+        _is_vec = channel_dim is not None
+        if _is_vec:
+            data_array = np.moveaxis(data_array, -1, 0)  # from channel last to channel first
```
I think here you only revert the order of spatial-dims for ITK, it's not consistent with our `ITKReader` logic for **single channel data**, which reverts all the dims:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/image_reader.py#L346
Could you please help make them consistent in this PR?

Thanks.

wyli replied:
they are consistent here, it is in https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/image_reader.py#L348, there will be unit tests to check  `SaveImage`/`LoadImage` in the follow-up PRs.

Nic-Ma replied:
I mean for **single channel** data, would it be `img.GetNumberOfComponentsPerPixel() == 1`? So the logic will be different from here.

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        if _is_vec:
+            data_array = np.moveaxis(data_array, -1, 0)  # from channel last to channel first
+        data_array = data_array.T.astype(dtype, copy=True, order="C")
+        itk_obj = itk.GetImageFromArray(data_array, is_vector=_is_vec, ttype=kwargs.pop("ttype", None))
```
May I know the usage case of this `is_vector` arg here?
I don't quite understand the relationship between `is_vec` and ITK dim order vs numpy dim order.

Thanks in advance.

wyli replied:
this flag and the dtype will determine the underlying itk object type, for example
```py
>>> import numpy as np
>>> array = np.zeros((1,2,3))
>>> import itk
>>> itk.GetImageFromArray(array, is_vector=False)
<itk.itkImagePython.itkImageD3; proxy of <Swig Object of type 'itkImageD3 *' at 0x7f8b5aedb600> >
>>> itk.GetImageFromArray(array, is_vector=True)
<itk.itkVectorImagePython.itkVectorImageD2; proxy of <Swig Object of type 'itkVectorImageD2 *' at 0x7f8b68fad1e0> >
```
the same array could be interpreted as a 3d spatial image or a 2d vector image.

Nic-Ma replied:
I still don't understand this line: `_is_vec = channel_dim is not None`, why if `channel_dim` is not None, then the data is **2d vector image**?

Thanks.

wyli replied:
in the example `np.zeros((1,2,3))` could be saved as a 2d image with channels (e.g. `channel_dim=-1` for channel-last) or a 3d image without any channel (`channel_dim=None`).

Nic-Ma replied:
But if the data is 4d like `[128, 128, 128, 3]` and `channel_dim` is `-1`, it will be treated as `_is_vec = True`?

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+    """
+
+    def __init__(self, output_dtype: DtypeLike = np.float32, **kwargs):
+        super().__init__(output_dtype=output_dtype, affine=None, channel_dim=0, **kwargs)
```
Can you also put `affine` and `channel_dim`  in the `__init__` of ITKWriter?

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        if verbose:
+            logger.info(f"writing: {filename_or_obj}")
+
+    @classmethod
```
May I know why using `classmethod` here? Maybe making it into a utility function directly?

Thanks.

wyli replied:
subclasses can extend this method for different backend, and they share some common logic such as casting torch tensor to ndarray.

Nic-Ma replied:
OK, I see. Seems you don't want to implement it as `static` function and it also works with other class functions.

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        itk_obj.SetDirection(itk.GetMatrixFromArray(_direction))
+        return itk_obj
+
+    @staticmethod
```
May I know why using `classmethod` here? Maybe making it into a utility function directly?

Thanks.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+            data = ascontiguousarray(data)
+        return data
+
+    @classmethod
```
May I know why using `classmethod` here? Maybe making it into a utility function directly?

Thanks.

wyli replied:
the PIL writer would override this class as affine is not needed in that case

Nic-Ma reviewed (commented):
Thanks for the quick implementation.
The overall logic looks good to me, except for the `resample`, `as_channel_last` logic.
Put some minor comments inline.

Thanks.

wyli commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        sp_dims = min(data.ndim, 3)
+        output_spatial_shape_ += [1] * (sp_dims - len(output_spatial_shape_))
+        output_spatial_shape_ = output_spatial_shape_[:sp_dims]
+        original_channels = data.shape[3:]
```
for spatial dimensions, it supports up to 3D, and it could have multiple time/modality dimensions.

Nic-Ma replied:
I mean `[3:]` can't get the channel dim of 2D data, maybe use `sp_dims` instead?

Thanks.

wyli replied:
maybe the docstring answers your question, it follows the nifti format as described in the docstring https://github.com/Project-MONAI/MONAI/blob/20d0bf93b348ef25f8dad066869233302d8e293d/monai/data/image_writer.py#L151-L159 

I'll try to make it a generic function/transform, it'll be useful for the metadata handling.

ericspod commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+        return convert_data_type(data_array, np.ndarray)[0]  # type: ignore
+
+    @classmethod
+    def resample_if_needed(
```
It's probably better to separate functionality so that definitions do only one thing at a time unless there's good reason to do multiples. I guess that's what the feedback was saying?

wyli replied:
thanks this is addressed in https://github.com/Project-MONAI/MONAI/pull/3701.

Nic-Ma replied:
Cool! thanks for the quick update!

Nic-Ma commented on monai/data/utils.py:
```diff
@@ -547,6 +548,30 @@ def set_rnd(obj, seed: int) -> int:
     return seed
 
 
+def affine_to_spacing(affine: NdarrayTensor, dims: int = 3, dtype=float, suppress_zeros: bool = True) -> NdarrayTensor:
```
Hi @wyli ,

I think this is the function used in: https://github.com/Project-MONAI/MONAI/issues/3766, right?

Thanks.

wyli replied:
Yes, this commit  is to address it,  commit message links to it https://github.com/wyli/MONAI/commit/b32336e596009127c46c813149296fa835697471

Nic-Ma commented on monai/data/utils.py:
```diff
@@ -547,6 +548,30 @@ def set_rnd(obj, seed: int) -> int:
+
+    Args:
+        affine: a d x d affine matrix.
+        dims: indexing of the spatial dimensions `affine[:dims, :dims]`.
```
Maybe we should use arg name `ndim` to be consistent with others?

Thanks.

wyli replied:
it's not actually ndim, it could be `-1` here for the indexing.

Nic-Ma replied:
If so, I would suggest to change another arg name, `dims` looks like a list of something, confusing.

Thanks.

wyli replied:
sure updated to `r`.

Nic-Ma commented on monai/data/utils.py:
```diff
@@ -60,6 +60,7 @@
     "list_data_collate",
     "worker_init_fn",
     "set_rnd",
+    "affine_to_spacing",
```
Please also add new function `orientation_ras_lps` here.

Thanks.

Nic-Ma reviewed (approved):
Thanks for the quick update.
Looks good to me now, except for several minor comments inline.
@ericspod @rijobro Do you have any other comments?

Thanks.

wyli commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,362 @@
+            spatial_ndim: modifying the spatial dims if needed, so that output to have at least
+                this number of spatial dims. If ``None``, the output will have the same number of
+                spatial dimensions as the input.
+            contiguous: if ``True``, the output will be contiguous.
```
it'll cause a copying operation, it's mostly not in use.

wyli commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,362 @@
+        data_array = super().create_backend_obj(data_array)
+        _is_vec = channel_dim is not None
+        if _is_vec:
+            data_array = np.moveaxis(data_array, -1, 0)  # from channel last to channel first
```
not really, after `set_data_array` the `data_array` here is channel-last.

wyli commented:
/black

wyli commented:
> Thanks for the quick update. Looks good to me now, except for several minor comments inline.

thanks Nic, I updated the PR for this use case of writing with a different spatial size:
```py
# write as 2d image, channel-first
np_data = np.arange(48).reshape(3, 4, 4)
writer = ITKWriter(output_dtype=np.uint8)
writer.set_data_array(np_data, channel_dim=0)
writer.set_metadata({"spatial_shape": (5, 5)})
writer.write("test1.png")
```

could you review again to confirm?
also when writing the examples I found that
```py
writer = ITKWriter(channel_dim=0)
writer.set_data_array(..., channel_dim=0)
```
will cause confusion, so I removed the constructor's `channel_dim`.

Nic-Ma commented on monai/data/image_writer.py:
```diff
@@ -0,0 +1,395 @@
+    def write(self, filename: PathLike, verbose: bool = True, **kwargs):
+        """subclass should implement this method to call the backend-specific writing APIs."""
+        if verbose:
+            logger.info(f"writing: {filename}")
```
Seems the `logger` is not defined? Could you please help confirm it?

Thanks.

wyli replied:
it's a module-level variable, now it's added to `__all__`

Nic-Ma reviewed (approved):
Looks good to me except for 2 minor comments inline.

Thanks.

wyli commented:
/build

wyli commented:
/build
